controller/src/main/java/io/pravega/controller/server/rentention/AutoRetentionBucketService.java

```diff
@@ -0,0 +1,146 @@
+        this.bucketId = bucketId;
+        this.streamMetadataStore = streamMetadataStore;
+        this.streamMetadataTasks = streamMetadataTasks;
+        this.executor = executor;
+
+        this.notifications = new LinkedBlockingQueue<>();
+        this.stop = new AtomicBoolean(false);
+        this.retentionFutureMap = new ConcurrentHashMap<>();
+    }
+
+    @Override
+    protected void doStart() {
+        RetryHelper.withIndefiniteRetriesAsync(() -> streamMetadataStore.getStreamsForBucket(bucketId, executor)
+                .thenAccept(streams -> retentionFutureMap.putAll(streams.stream()
+                        .map(s -> {
+                            String[] splits = s.split("/");
+                            return new StreamImpl(splits[0], splits[1]);
+                        })
+                        .collect(Collectors.toMap(s -> s, this::getStreamRetentionFuture))
+                )),
+                e -> log.warn("exception thrown getting streams for bucket {}, e = {}", bucketId, e), executor)
+                .thenAccept(x -> {
+                    log.info("streams collected for the bucket, registering for change notification and starting loop for processing notifications");
+                    streamMetadataStore.registerBucketChangeListener(bucketId, this);
+                    Futures.loop(stop::get, this::processNotification, executor);
+                })
+                .whenComplete((r, e) -> {
+                    if (e != null) {
+                        notifyFailed(e);
+                    } else {
+                        notifyStarted();
+                    }
+                });
+    }
+
+    private CompletableFuture<Void> processNotification() {
+        return CompletableFuture.runAsync(() -> {
+            StreamNotification notification = Exceptions.handleInterrupted(notifications::take);
+            StreamImpl stream = new StreamImpl(notification.getScope(), notification.getStream());
+
+            switch (notification.getType()) {
+                case StreamAdded:
+                    retentionFutureMap.putIfAbsent(stream, getStreamRetentionFuture(stream));
+                    break;
+                case StreamRemoved:
+                    retentionFutureMap.remove(stream).cancel(true);
+                    break;
+                case StreamUpdated:
+                    // if retention has been disabled then cancel the future.
+                    // if retention has been enabled then add the future if it does not exist.
+                    // For now we will do nothing.
+                    break;
+                case ConnectivityError:
+                    // TODO: refresh retention future map as this means we can miss events
+                    break;
+            }
+        });
+    }
+
+    private CompletableFuture<Void> getStreamRetentionFuture(StreamImpl stream) {
+        return Futures.loopWithDelay(stop::get, () -> performRetention(stream),
```

Comment:
Why can't we use `isRunning` here like @andreipaduroiu suggested above?

Reply:
We can, however, it races with start. 
When we start, the state is set to running as we start this loop. 
We can use the state as !stopping && !stopped.